core/src/main/java/com/linecorp/armeria/client/EventLoopScheduler.java

```diff
@@ -0,0 +1,322 @@
+ * under the License.
+ */
+
+package com.linecorp.armeria.client;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ThreadLocalRandom;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.Joiner;
+import com.google.common.collect.Streams;
+
+import com.linecorp.armeria.common.util.Releasable;
+
+import io.netty.channel.EventLoop;
+import io.netty.channel.EventLoopGroup;
+
+final class EventLoopScheduler {
+
+    private final EventLoop[] eventLoops;
+    private final Map<String, State> map = new ConcurrentHashMap<>();
+    private int counter;
+    private volatile long lastCleanupTimeNanos = System.nanoTime();
+
+    EventLoopScheduler(EventLoopGroup eventLoopGroup) {
+        eventLoops = Streams.stream(eventLoopGroup)
+                            .map(e -> (EventLoop) e)
+                            .toArray(EventLoop[]::new);
+    }
+
+    Entry acquire(Endpoint endpoint) {
+        requireNonNull(endpoint, "endpoint");
+        final State state = state(endpoint);
+        final Entry acquired = state.acquire();
+        cleanup();
+        return acquired;
+    }
+
+    @VisibleForTesting
+    List<Entry> entries(Endpoint endpoint) {
+        return state(endpoint).entries();
+    }
+
+    private State state(Endpoint endpoint) {
+        final String authority = endpoint.authority();
+        return map.computeIfAbsent(authority, e -> new State(eventLoops));
+    }
+
+    /**
+     * Cleans up empty entries with no activity for more than 1 minute. For reduced overhead, we perform this
+     * only when 1) the last clean-up was more than 1 minute ago and 2) the number of acquisitions % 256 is 0.
+     */
+    private void cleanup() {
+        if ((++counter & 0xFF) != 0) {
+            return;
+        }
+
+        final long currentTimeNanos = System.nanoTime();
+        if (currentTimeNanos - lastCleanupTimeNanos < 60_000_000_000L) {
+            return;
+        }
+
+        for (Iterator<State> i = map.values().iterator(); i.hasNext();) {
+            final State state = i.next();
+            final boolean remove;
+
+            synchronized (state) {
+                remove = state.allActiveRequests == 0 &&
+                         currentTimeNanos - state.lastActivityTimeMillis >= 60_000_000_000L;
+            }
+
+            if (remove) {
+                i.remove();
+            }
+        }
+
+        lastCleanupTimeNanos = System.nanoTime();
+    }
+
+    private static final class State {
+        /**
+         * A binary heap of Entry. Ordered by:
+         * <ul>
+         *   <li>{@link Entry#activeRequests()} (lower is better)</li>
+         *   <li>{@link Entry#id()} (lower is better)</li>
+         * </ul>
+         */
+        private final List<Entry> entries;
```

Comment:
I can update the priority of an element 1) at arbitrary index 2) without removing and adding it back.